src/dynamic_batch_scheduler.cc

```diff
@@ -358,7 +358,20 @@ DynamicBatchScheduler::BatcherThread(const int nice)
         if (payload_saturated_) {
           continue;
         }
-        cv_.wait(lock, wait_for_slots);
+
+        {
+          // The wait_for_slots conditional can be blocking till the slots
+          // are available for execution. Need to explicitly release the
+          // outer lock to allow Enqueue threads above to make progress.
+          lock.unlock();
+          // Use slot lock to wait for the slot availability.
+          std::unique_lock<std::mutex> slot_lock(slot_mu_);
+          cv_.wait(slot_lock, wait_for_slots);
```

Comment:
As discussed offline, because `wait_for_slots()` is blocking until payload slot is available, I think we should remove the lock/cv usage here, as it introduces some confusion. There is no other threads here trying to acquire slot_lock/slot_mu

This chunk should basically:
```cpp
unlock()
block_until_slot/consumer_available
lock()
```

Reply:
In our discussion today, we missed the case where the queue supports pre-fetching. The payload callback would need to notify the batcher thread in this case as soon as the slot is made available. I think it is best to keep a unified interface for both the variants.

Reply:
Makes sense. Do you think the comments here or some other kind of docs could better summarize the high level process involved here between the handler <-> batcher <-> backend threads for future readers that we spent a while talking about?

Reply:
One potential here is a "retrospective design doc" that would describe an area in detail

Reply:
Created DLIS-5785 to write a architecture doc for the current implementation of dynamic batcher and the thread interactions there in.